clang/test/Driver/clang-offload-wrapper-exe-preview.cpp

```diff
@@ -0,0 +1,212 @@
+  void *ValAddr;    // address of property value
+  uint32_t Type;    // pi_property_type
+  uint64_t ValSize; // size of property value in bytes
+};
+
+typedef _pi_device_binary_property_struct *pi_device_binary_property;
+
+struct _pi_device_binary_property_set_struct {
+  char *Name;                                // the name
+  pi_device_binary_property PropertiesBegin; // array start
+  pi_device_binary_property PropertiesEnd;   // array end
+};
+
+typedef _pi_device_binary_property_set_struct *pi_device_binary_property_set;
+
+struct pi_device_binary_struct {
+  uint16_t Version;
+  uint8_t Kind;   // 4 for SYCL
+  uint8_t Format; // 1 for native
+  const char *DeviceTargetSpec;
+  const char *CompileOptions;
+  const char *LinkOptions;
+  const unsigned char *BinaryStart;
+  const unsigned char *BinaryEnd;
+  _pi_offload_entry EntriesBegin;
+  _pi_offload_entry EntriesEnd;
+  pi_device_binary_property_set PropertySetsBegin;
+  pi_device_binary_property_set PropertySetsEnd;
+};
+typedef pi_device_binary_struct *pi_device_binary;
+
+struct pi_device_binaries_struct {
+  uint16_t Version;
+  uint16_t NumDeviceBinaries;
+  pi_device_binary DeviceBinaries;
+  _pi_offload_entry *HostEntriesBegin;
+  _pi_offload_entry *HostEntriesEnd;
+};
+typedef pi_device_binaries_struct *pi_device_binaries;
+
+static pi_device_binaries BinDesc = nullptr;
+
+// Wrapper object has code which calls these 2 functions below
+extern "C" void __sycl_register_lib(pi_device_binaries desc) {
+  BinDesc = desc;
+}
+
+extern "C" void __sycl_unregister_lib() {}
+
+#define ASSERT(Cond, Msg)                            \
+  if (!(Cond)) {                                     \
+    std::cerr << "*** ERROR: wrong " << Msg << "\n"; \
+    return 1;                                        \
+  }
+
+static std::string getString(const unsigned char *B, const unsigned char *E) {
+  return std::string(reinterpret_cast<const char *>(B), E - B);
+}
+
+static int getInt(void *Addr) {
+  const char *Ptr = reinterpret_cast<const char *>(Addr);
+  return Ptr[0] | (Ptr[1] << 8) | (Ptr[2] << 16) | (Ptr[3] << 24);
+}
+
+using byte = unsigned char;
+
+static void printProp(const pi_device_binary_property &Prop) {
+  std::cerr << "Property " << Prop->Name << " {\n";
+  std::cerr << "  Type: " << Prop->Type << "\n";
+  if (Prop->Type != 1)
+    std::cerr << "  Size = " << Prop->ValSize << "\n";
+
+  std::cerr << "  Value = ";
+  if (Prop->Type == 1)
+    std::cerr << getInt(&Prop->ValSize);
+  else {
+    std::cerr << " {\n   ";
+
+    byte *Ptr = (byte *)Prop->ValAddr;
+
+    for (auto I = 0; I < Prop->ValSize && I < 100; ++I) {
+      std::cerr << " 0x" << std::hex << (unsigned int)Ptr[I];
+      std::cerr << std::dec;
+    }
+    std::cerr << "\n  }";
+  }
+  std::cerr << "\n";
+  std::cerr << "}\n";
+}
+
+static int dumpBinary0() {
+  pi_device_binary Bin = &BinDesc->DeviceBinaries[0];
+  ASSERT(Bin->Kind == 4, "Bin->Kind");
+  ASSERT(Bin->Format == 1, "Bin->Format");
+
+  // dump code
```

Comment:
This is a nit, but according to LLVM Coding Standards, we need to use proper English for comments, including upper-case at the beginning of sentences, and proper punctuation. There are a few more comments that would need to be updated.
```suggestion
  // Dump code.
```